### DISCUSSION TITLE
Bluetooth with DSP and Internal DAC

### DISCUSSION THREAD

**Discussion by InsaneEngineer** [pschatzmann/ESP32-A2DP#275] [Q&A] [answered]
I want to some research limits of ESP32 in audio processing. But i dont have external DAC.  I am override a callback function, but original code of this function very hard to understand:
```
                if (is_i2s_output) {
		// special case for internal DAC output, the incomming PCM buffer needs
		// to be converted from signed 16bit to unsigned
		int16_t* data16 = (int16_t*)data;

		if (this->i2s_config.mode & I2S_MODE_DAC_BUILT_IN) {
			//HACK: this is here to remove the const restriction to replace the data in place as per
			//https://github.com/espressif/esp-idf/blob/178b122/components/bt/host/bluedroid/api/include/api/esp_a2dp_api.h
			//the buffer is anyway static block of memory possibly overwritten by next incomming data.

			for (int i = 0; i < len / 2; i++) {
				int16_t sample = data[i * 2] | data[i * 2 + 1] << 8;
				data16[i] = sample + 0x8000;
			}
		}
```

P.s. I had some knowledge in therory of Digital Filters and my main object - it iss creating some sort of equalizer. For first attempt i am creating a simple digital low pass filter:
![image](https://user-images.githubusercontent.com/64024820/183256315-7b078fa4-8318-4cf7-9020-88de0d13ccc1.png)

**Comment by pschatzmann** [comment, marked as answer]
I suggest that you have a look at my [AudioTools](https://github.com/pschatzmann/arduino-audio-tools/) library where you don't need to bother about these details. 
Here is an example using [an Equilizer](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/basic-a2dp-eq-audiokit/basic-a2dp-eq-audiokit.ino). Just change the AudioKitStream to an AnalogAudioStream to use the internal DAC.

If you use the internal DAC with the ESP-IDF API,  bits_per_sample is 16 but the DAC module will only take the 8bits from MSB.
